pkg/api/merge.go

```diff
@@ -0,0 +1,368 @@
+		return nil
+	case complete == nil:
+		complete = &OpenShiftManagedCluster{}
+	case partial == nil:
+		partial = &OpenShiftManagedCluster{}
+	}
+
+	return &OpenShiftManagedCluster{
+		ID:         mergeString(complete.ID, partial.ID),
+		Location:   mergeString(complete.Location, partial.Location),
+		Name:       mergeString(complete.Name, partial.Name),
+		Plan:       mergeResourcePurchasePlan(complete.Plan, partial.Plan),
+		Tags:       mergeTags(complete.Tags, partial.Tags),
+		Type:       mergeString(complete.Type, partial.Type),
+		Properties: mergeProperties(complete.Properties, partial.Properties),
+		Config:     complete.Config, // NOTE: Config isn't merged
+	}
+}
+
+func mergeString(complete, partial string) string {
+	if partial == "" {
+		return complete
+	}
+	return partial
+}
+
+func mergeResourcePurchasePlan(complete, partial *ResourcePurchasePlan) *ResourcePurchasePlan {
+	switch {
+	case complete == nil && partial == nil:
+		return nil
+	case complete == nil:
+		complete = &ResourcePurchasePlan{}
+	case partial == nil:
+		partial = &ResourcePurchasePlan{}
+	}
+
+	return &ResourcePurchasePlan{
+		Name:          mergeString(complete.Name, partial.Name),
+		Product:       mergeString(complete.Product, partial.Product),
+		PromotionCode: mergeString(complete.PromotionCode, partial.PromotionCode),
+		Publisher:     mergeString(complete.Publisher, partial.Publisher),
+	}
+}
+
+func mergeTags(complete, partial map[string]string) map[string]string {
+	if complete == nil && partial == nil {
+		return nil
+	}
+
+	// TODO: is this the semantic that we want for mergeTags?
+
+	rv := make(map[string]string, len(complete))
+	for k, v := range complete {
+		rv[k] = v
+	}
+	for k, v := range partial {
+		rv[k] = v
+	}
+	return rv
+}
+
+func mergeProperties(complete, partial *Properties) *Properties {
+	switch {
+	case complete == nil && partial == nil:
+		return nil
+	case complete == nil:
+		complete = &Properties{}
+	case partial == nil:
+		partial = &Properties{}
+	}
+
+	return &Properties{
+		ProvisioningState:       mergeProvisioningState(complete.ProvisioningState, partial.ProvisioningState),
+		OpenShiftVersion:        mergeString(complete.OpenShiftVersion, partial.OpenShiftVersion),
+		PublicHostname:          mergeString(complete.PublicHostname, partial.PublicHostname),
+		FQDN:                    mergeString(complete.FQDN, partial.FQDN),
+		RouterProfiles:          mergeRouterProfiles(complete.RouterProfiles, partial.RouterProfiles),
+		AgentPoolProfiles:       mergeAgentPoolProfiles(complete.AgentPoolProfiles, partial.AgentPoolProfiles),
+		AuthProfile:             mergeAuthProfile(complete.AuthProfile, partial.AuthProfile),
+		ServicePrincipalProfile: mergeServicePrincipalProfile(complete.ServicePrincipalProfile, partial.ServicePrincipalProfile),
+		AzProfile:               mergeAzProfile(complete.AzProfile, partial.AzProfile),
+	}
+}
+
+func mergeProvisioningState(complete, partial ProvisioningState) ProvisioningState {
```

Comment:
nit: these are just strings we could reuse `mergeString` (thought it might look a bit messy since casting which is why you may not have)